presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ConvertJoinTreeToJoinGraph.java

```diff
@@ -19,58 +19,37 @@
 import com.facebook.presto.sql.planner.SymbolAllocator;
 import com.facebook.presto.sql.planner.iterative.Lookup;
 import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.Assignments;
 import com.facebook.presto.sql.planner.plan.JoinNode;
 import com.facebook.presto.sql.planner.plan.PlanNode;
-import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
 
-import java.util.ArrayList;
-import java.util.List;
 import java.util.Optional;
 
-import static com.facebook.presto.sql.ExpressionUtils.extractConjuncts;
+import static com.facebook.presto.SystemSessionProperties.isJoinReorderingEnabled;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
 
 public class ConvertJoinTreeToJoinGraph
         implements Rule
 {
     @Override
     public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
     {
-        // We check that join distribution type is absent because we only want to do this transformation once.
-        if (node instanceof JoinNode && ((JoinNode) node).getType().equals(JoinNode.Type.INNER) && !((JoinNode) node).getDistributionType().isPresent()) {
-            return Optional.of(new FlattenedJoinTree((JoinNode) node).toJoinGraphNode(idAllocator));
-        }
-        else {
+        // We check that join distribution type is absent because we only want to do this transformation once (reordered joins will have distribution type already set).
+        if (!(node instanceof JoinNode) || !isJoinReorderingEnabled(session)) {
             return Optional.empty();
         }
-    }
-
-    private static class FlattenedJoinTree
-    {
-        private final List<PlanNode> nodes = new ArrayList<>();
-        private final List<JoinNode.EquiJoinClause> criteria = new ArrayList<>();
-        private final List<Expression> filters = new ArrayList<>();
-
-        private FlattenedJoinTree(JoinNode node)
-        {
-            flattenNode(node);
-        }
 
-        private void flattenNode(PlanNode node)
-        {
-            if (node instanceof JoinNode && ((JoinNode) node).getType() == JoinNode.Type.INNER) {
-                flattenNode(((JoinNode) node).getLeft());
-                flattenNode(((JoinNode) node).getRight());
-                criteria.addAll(((JoinNode) node).getCriteria());
-                ((JoinNode) node).getFilter().ifPresent(f -> filters.addAll(extractConjuncts(f)));
-            }
-            else {
-                nodes.add(node);
+        JoinNode joinNode = (JoinNode) node;
+        if (joinNode.getType().equals(INNER) && !joinNode.getDistributionType().isPresent()) {
```

Comment:
I would make it the other way around. Move the code path that returns `Optional.empty()` inside `if`. The rewrite code path (let's call it default) should be top level